blob/service.go

```diff
@@ -354,14 +354,153 @@ func (s *Service) retrieve(
 		return nil, nil, err
 	}
-	headerGetterSpan.SetStatus(codes.Ok, "")
-	headerGetterSpan.AddEvent("received eds", trace.WithAttributes(
+
+	getSharesSpan.SetStatus(codes.Ok, "")
+	getSharesSpan.AddEvent("received shares", trace.WithAttributes(
 		attribute.Int64("eds-size", int64(len(header.DAH.RowRoots)))))
 
+	var (
+		appShares = make([]shares.Share, 0)
+		proofs    = make(namespaceToRowRootProof, 0)
+	)
+
+	for _, row := range namespacedShares {
+		if len(row.Shares) == 0 {
+			// the above condition means that we've faced with an Absence Proof.
+			// This Proof proves that the namespace was not found in the DAH, so
+			// we can return `ErrBlobNotFound`.
+			return nil, nil, ErrBlobNotFound
+		}
+
+		appShares, err = toAppShares(row.Shares...)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		proofs = append(proofs, row.Proof)
+		index := row.Proof.Start()
+
+		for {
+			var (
+				isComplete bool
+				shrs       []shares.Share
+				wasEmpty   = sharesParser.isEmpty()
+			)
+
+			if wasEmpty {
+				// create a parser if it is empty
+				shrs, err = sharesParser.set(rowIndex*len(header.DAH.RowRoots)+index, appShares)
+				if err != nil {
+					if errors.Is(err, errEmptyShares) {
+						// reset parser as `skipPadding` can update next blob's index
+						sharesParser.reset()
+						appShares = nil
+						break
+					}
+					return nil, nil, err
+				}
+
+				// update index and shares if padding shares were detected.
+				if len(appShares) != len(shrs) {
+					index += len(appShares) - len(shrs)
+					appShares = shrs
+				}
+			}
+
+			shrs, isComplete = sharesParser.addShares(appShares)
+			// move to the next row if the blob is incomplete
+			if !isComplete {
+				appShares = nil
+				break
+			}
+			// otherwise construct blob
+			blob, err := sharesParser.parse()
+			if err != nil {
+				return nil, nil, err
+			}
+
+			if sharesParser.verify(blob) {
+				return blob, &proofs, nil
+			}
+
+			index += len(appShares) - len(shrs)
+			appShares = shrs
+			sharesParser.reset()
+
+			if !wasEmpty {
+				// remove proofs for prev rows if verified blob spans multiple rows
+				proofs = proofs[len(proofs)-1:]
+			}
+		}
+
+		rowIndex++
+		if sharesParser.isEmpty() {
+			proofs = nil
+		}
+	}
+
+	err = ErrBlobNotFound
+	for _, sh := range appShares {
+		ok, err := sh.IsPadding()
+		if err != nil {
+			return nil, nil, err
+		}
+		if !ok {
+			err = fmt.Errorf("incomplete blob with the "+
+				"namespace: %s detected at %d: %w", namespace.String(), height, err)
+			log.Error(err)
+		}
+	}
+	return nil, nil, err
+}
+
+// retrieve retrieves blobs and their proofs by requesting the whole namespace and
+// comparing Commitments.
+// Retrieving is stopped once the `verify` condition in shareParser is met.
+func (s *Service) retrieveBlobProof(
+	ctx context.Context,
+	height uint64,
+	namespace share.Namespace,
+	sharesParser *parser,
+) (_ *Blob, _ *Proof, err error) {
+	log.Infow("requesting blob",
```

Comment:
```suggestion
	log.Infow("requesting blob proof",
```

because proof is now retrieved via `retrieve`